backfill/src/main.rs

```diff
@@ -0,0 +1,161 @@
+use std::ops::RangeInclusive;
+
+use async_trait::async_trait;
+use futures::TryStreamExt;
+use jsonrpsee::{
+    core::RpcResult,
+    proc_macros::rpc,
+    types::{error::INTERNAL_ERROR_CODE, ErrorObject},
+};
+use reth::primitives::{BlockNumber, Requests};
+use reth_execution_types::{Chain, ExecutionOutcome};
+use reth_exex::{BackfillJobFactory, ExExContext, ExExEvent, ExExNotification};
+use reth_node_api::FullNodeComponents;
+use reth_node_ethereum::EthereumNode;
+use reth_tracing::tracing::{error, info};
+use tokio::sync::mpsc;
+
+/// The ExEx that consumes new [`ExExNotification`]s and processes new backfill requests by
+/// [`BackfillRpcExt`].
+struct BackfillExEx<Node: FullNodeComponents> {
+    /// The context of the ExEx.
+    ctx: ExExContext<Node>,
+    /// Receiver for backfill requests.
+    backfill_rx: mpsc::UnboundedReceiver<RangeInclusive<BlockNumber>>,
+    /// Factory for backfill jobs.
+    backfill_job_factory: BackfillJobFactory<Node::Executor, Node::Provider>,
+}
+
+impl<Node: FullNodeComponents> BackfillExEx<Node> {
+    /// Creates a new instance of the ExEx.
+    fn new(
+        ctx: ExExContext<Node>,
+        backfill_rx: mpsc::UnboundedReceiver<RangeInclusive<BlockNumber>>,
+    ) -> Self {
+        let backfill_job_factory =
+            BackfillJobFactory::new(ctx.block_executor().clone(), ctx.provider().clone());
+        Self { ctx, backfill_rx, backfill_job_factory }
+    }
+
+    /// Starts listening for notifications and backfill requests.
+    async fn start(mut self) -> eyre::Result<()> {
+        loop {
+            tokio::select! {
+                Some(notification) = self.ctx.notifications.recv() => {
+                    self.process_notification(notification).await?;
+                }
+                Some(range) = self.backfill_rx.recv() => {
+                    let _ = self.backfill(range).await.inspect_err(|err| error!(%err, "Backfill error occurred"));
```

Comment:
this will effectively wait until the range is done, and could starve notification processing, we should just spawn this, ideally we also spawn and use some semaphor as guard so we limit the concurrent backfill requests